src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking.cs

```diff
@@ -19415,6 +19415,176 @@ public void ExplicitCast_StaticType()
                 Diagnostic(ErrorCode.ERR_ConvertToStaticClass, "(C?)y").WithArguments("C").WithLocation(4, 36));
         }
 
+        [Fact]
+        public void ForEach_01()
+        {
+            var source =
+@"class Enumerable
+{
+    public Enumerator GetEnumerator() => new Enumerator();
+}
+class Enumerator
+{
+    public object Current => throw null;
+    public bool MoveNext() => false;
+}
+class C
+{
+    static void F(Enumerable e)
+    {
+        foreach (var x in e)
+            x.ToString();
+        foreach (string y in e)
+            y.ToString();
+        foreach (string? z in e)
+            z.ToString();
+    }
+}";
+            var comp = CreateCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics();
```

Comment:
>comp.VerifyDiagnostics(); [](start = 12, length = 25)

Should probably warn in all three dereference cases once the APIs are properly annotated.

Reply:
The test should fail at that point.

---
In reply to: [186531353](https://github.com/dotnet/roslyn/pull/26660#discussion_r186531353) [](ancestors = 186531353)

Reply:
> The test should fail at that point.

It would be good to leave a comment that this is what we expect to happen eventually.

---
In reply to: [186746097](https://github.com/dotnet/roslyn/pull/26660#discussion_r186746097) [](ancestors = 186746097,186531353)

Reply:
Added comment to `ForEach_04`.

---
In reply to: [186810274](https://github.com/dotnet/roslyn/pull/26660#discussion_r186810274) [](ancestors = 186810274,186746097,186531353)